File: libraries/AdaptiveExpressions/parser/ExpressionParser.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Text.RegularExpressions;
using AdaptiveExpressions.parser;
using Antlr4.Runtime;
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Tree;
using Newtonsoft.Json.Linq;

[assembly: CLSCompliant(false)]

namespace AdaptiveExpressions
{
    /// <summary>
    /// Parser to turn strings into an <see cref="Expression"/>.
    /// </summary>
    public class ExpressionParser : IExpressionParser
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ExpressionParser"/> class.
        /// Constructor.
        /// </summary>
        /// <param name="lookup">Delegate to lookup evaluation information from type string.</param>
        public ExpressionParser(EvaluatorLookup lookup = null)
        {
            EvaluatorLookup = lookup ?? Expression.Lookup;
        }

        /// <summary>
        /// Gets the elegate to lookup function information from the type.
        /// </summary>
        /// <value>
        /// The elegate to lookup function information from the type.
        /// </value>
        public EvaluatorLookup EvaluatorLookup { get; }

        /// <summary>
        /// Parse the input into an expression.
        /// </summary>
        /// <param name="expression">Expression to parse.</param>
        /// <returns>Expression tree.</returns>
        public Expression Parse(string expression)
        {
            if (string.IsNullOrEmpty(expression))
            {
                return Expression.ConstantExpression(string.Empty);
            }
            else
            {
                return new ExpressionTransformer(EvaluatorLookup).Transform(AntlrParse(expression));
            }
        }

        protected static IParseTree AntlrParse(string expression)
        {
            var inputStream = new AntlrInputStream(expression);
            var lexer = new ExpressionAntlrLexer(inputStream);
            lexer.RemoveErrorListeners();
            var tokenStream = new CommonTokenStream(lexer);
            var parser = new ExpressionAntlrParser(tokenStream);
            parser.RemoveErrorListeners();
            parser.AddErrorListener(ParserErrorListener.Instance);
            parser.BuildParseTree = true;
            return parser.file()?.expression();
        }

        private class ExpressionTransformer : ExpressionAntlrParserBaseVisitor<Expression>
        {
            private readonly Regex escapeRegex = new Regex(@"\\[^\r\n]?");
            private readonly EvaluatorLookup _lookupFunction;

            public ExpressionTransformer(EvaluatorLookup lookup)
            {
                _lookupFunction = lookup;
            }

            public Expression Transform(IParseTree context) => Visit(context);

            public override Expression VisitUnaryOpExp([NotNull] ExpressionAntlrParser.UnaryOpExpContext context)
            {
                var unaryOperationName = context.GetChild(0).GetText();
                var operand = Visit(context.expression());
                if (unaryOperationName == ExpressionType.Subtract
                    || unaryOperationName == ExpressionType.Add)
                {
                    return MakeExpression(unaryOperationName, new Constant(0), operand);
                }

                return MakeExpression(unaryOperationName, operand);
            }

            public override Expression VisitBinaryOpExp([NotNull] ExpressionAntlrParser.BinaryOpExpContext context)
            {
                var binaryOperationName = context.GetChild(1).GetText();
                var left = Visit(context.expression(0));
                var right = Visit(context.expression(1));
                return MakeExpression(binaryOperationName, left, right);
            }

            public override Expression VisitFuncInvokeExp([NotNull] ExpressionAntlrParser.FuncInvokeExpContext context)
            {
                var parameters = ProcessArgsList(context.argsList()).ToList();

                // Remove the check to check primaryExpression is just an IDENTIFIER to support "." in template name
                var functionName = context.primaryExpression().GetText();
                if (context.NON() != null)
                {
                    functionName += context.NON().GetText();
                }

                return MakeExpression(functionName, parameters.ToArray());
            }

            public override Expression VisitIdAtom([NotNull] ExpressionAntlrParser.IdAtomContext context)
            {
                Expression result;
                var symbol = context.GetText();
                var normalized = symbol.ToLower();
                if (normalized == "false")
                {
                    result = Expression.ConstantExpression(false);
                }
                else if (normalized == "true")
                {
                    result = Expression.ConstantExpression(true);
                }
                else if (normalized == "null")
                {
                    result = Expression.ConstantExpression(null);
                }
                else
                {
                    result = MakeExpression(ExpressionType.Accessor, Expression.ConstantExpression(symbol));
                }

                return result;
            }

            public override Expression VisitIndexAccessExp([NotNull] ExpressionAntlrParser.IndexAccessExpContext context)
            {
                Expression instance;
                var property = Visit(context.expression());

                instance = Visit(context.primaryExpression());
                return MakeExpression(ExpressionType.Element, instance, property);
            }

            public override Expression VisitMemberAccessExp([NotNull] ExpressionAntlrParser.MemberAccessExpContext context)
            {
                var property = context.IDENTIFIER().GetText();
                var instance = Visit(context.primaryExpression());

                return MakeExpression(ExpressionType.Accessor, Expression.ConstantExpression(property), instance);
            }

            public override Expression VisitNumericAtom([NotNull] ExpressionAntlrParser.NumericAtomContext context)
            {
                if (int.TryParse(context.GetText(), out var intValue))
                {
                    return Expression.ConstantExpression(intValue);
                }

                if (double.TryParse(context.GetText(), NumberStyles.Any, CultureInfo.InvariantCulture, out var doubleValue))
                {
                    return Expression.ConstantExpression(doubleValue);
                }

                throw new Exception($"{context.GetText()} is not a number in expression '{context.GetText()}'");
            }

            public override Expression VisitParenthesisExp([NotNull] ExpressionAntlrParser.ParenthesisExpContext context) => Visit(context.expression());

            public override Expression VisitArrayCreationExp([NotNull] ExpressionAntlrParser.ArrayCreationExpContext context)
            {
                var parameters = ProcessArgsList(context.argsList()).ToList();
                return MakeExpression(ExpressionType.CreateArray, parameters.ToArray());
            }

            public override Expression VisitStringAtom([NotNull] ExpressionAntlrParser.StringAtomContext context)
            {
                var text = context.GetText();
                if (text.StartsWith("'") && text.EndsWith("'"))
                {
                    text = text.Substring(1, text.Length - 2).Replace("\\'", "'");
                }
                else if (text.StartsWith("\"") && text.EndsWith("\""))
                {
                    text = text.Substring(1, text.Length - 2).Replace("\\\"", "\"");
                }
                else
                {
                    throw new Exception($"Invalid string {text}");
                }

                return Expression.ConstantExpression(EvalEscape(text));
            }

            public override Expression VisitJsonCreationExp([NotNull] ExpressionAntlrParser.JsonCreationExpContext context)
            {
                var expr = this.MakeExpression(ExpressionType.Json, new Constant("{}"));
                if (context.keyValuePairList() != null)
                {
                    foreach (var kvPair in context.keyValuePairList().keyValuePair())
                    {
                        var key = string.Empty;
                        var keyNode = kvPair.key().children[0];
                        if (keyNode is ITerminalNode node)
                        {
                            if (node.Symbol.Type == ExpressionAntlrParser.IDENTIFIER)
                            {
                                key = node.GetText();
                            }
                            else
                            {
                                key = node.GetText().Substring(1, node.GetText().Length - 2);
                            }
                        }

                        expr = this.MakeExpression(ExpressionType.SetProperty, expr, new Constant(key), this.Visit(kvPair.expression()));
                    }
                }

                return expr;
            }

            public override Expression VisitStringInterpolationAtom([NotNull] ExpressionAntlrParser.StringInterpolationAtomContext context)
            {
                var children = new List<Expression>();
                foreach (var child in context.stringInterpolation().children)
                {
                    if (child is ITerminalNode node)
                    {
                        switch (node.Symbol.Type)
                        {
                            case ExpressionAntlrParser.TEMPLATE:
                                var expressionString = TrimExpression(node.GetText());
                                children.Add(Expression.Parse(expressionString, _lookupFunction));
                                break;
                            case ExpressionAntlrParser.ESCAPE_CHARACTER:
                                children.Add(Expression.ConstantExpression(EvalEscape(node.GetText().Replace("\\`", "`").Replace("\\$", "$"))));
                                break;
                            default:
                                break;
                        }
                    }
                    else
                    {
                        // text content
                        var text = EvalEscape(child.GetText());
                        children.Add(Expression.ConstantExpression(text));
                    }
                }

                return MakeExpression(ExpressionType.Concat, children.ToArray());
            }

            private Expression MakeExpression(string functionType, params Expression[] children)
                => Expression.MakeExpression(_lookupFunction(functionType) ?? throw new SyntaxErrorException($"{functionType} does not have an evaluator, it's not a built-in function or a custom function."), children);

            private IEnumerable<Expression> ProcessArgsList(ExpressionAntlrParser.ArgsListContext context)
            {
                if (context != null)
                {
                    foreach (var expression in context.expression())
                    {
                        yield return Visit(expression);
                    }
                }
            }

            private string EvalEscape(string text)
            {
                if (text == null)
                {
                    return string.Empty;
                }

                return escapeRegex.Replace(text, new MatchEvaluator(m =>
                {
                    var value = m.Value;
                    var commonEscapes = new List<string>() { "\\r", "\\n", "\\t", "\\\\" };
                    if (commonEscapes.Contains(value))
                    {
                        return Regex.Unescape(value);
                    }

                    return value;
                }));
            }

            private string TrimExpression(string expression)
            {
                var result = expression.Trim().TrimStart('$').Trim();

                if (result.StartsWith("{") && result.EndsWith("}"))
                {
                    result = result.Substring(1, result.Length - 2);
                }

                return result;
            }
        }
    }
}

File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/Parser/ExpressionParser.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "ExpressionParser.h"
#include "Expression.h"
#include "ExpressionType.h"
#include "Constant.h"

#include "ExpressionAntlrLexer.h"
#include "ExpressionAntlrParser.h"

#include <iostream>
#include <string>

ExpressionParser::ExpressionTransformer::ExpressionTransformer(EvaluatorLookup lookup)
{
    m_lookupFunction = lookup;
}

Expression* ExpressionParser::ExpressionTransformer::Transform(antlr4::tree::ParseTree* context)
{
    return visit(context);
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitFile(ExpressionAntlrParser::FileContext* ctx)
{
    std::vector<Expression*> expressions;

    auto unaryOperationName = ctx->children[0]->getText();
    auto operand = visit(ctx->expression());
    if (unaryOperationName == ExpressionType::Subtract
        || unaryOperationName == ExpressionType::Add)
    {
        expressions.push_back((new Constant(0)));
        expressions.push_back(operand);

        return MakeExpression(unaryOperationName, expressions);
    }

    // expressions.push_back(new Expression(operand));
    return MakeExpression(unaryOperationName, expressions);
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitStringAtom(ExpressionAntlrParser::StringAtomContext* ctx)
{
    std::string text = ctx->getText();
    /*
    if (text.StartsWith("'", StringComparison.Ordinal) && text.EndsWith("'", StringComparison.Ordinal))
    {
        text = text.Substring(1, text.Length - 2).Replace("\\'", "'");
    }
    else if (text.StartsWith("\"", StringComparison.Ordinal) && text.EndsWith("\"", StringComparison.Ordinal))
    {
        text = text.Substring(1, text.Length - 2).Replace("\\\"", "\"");
    }
    else
    {
        throw new Exception($"Invalid string {text}");
    }
    */

    return Expression::ConstantExpression(text);
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitUnaryOpExp(ExpressionAntlrParser::UnaryOpExpContext* ctx)
{
    auto unaryOperationName = ctx->children.at(0)->getText();
    auto operand = visit(ctx->expression());
    if (unaryOperationName == ExpressionType::Subtract
        || unaryOperationName == ExpressionType::Add)
    {
        return MakeExpression(unaryOperationName, std::vector<Expression*>{new Constant(0), operand});
    }

    return MakeExpression(unaryOperationName, std::vector<Expression*>{operand});
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitBinaryOpExp(ExpressionAntlrParser::BinaryOpExpContext* ctx)
{
    auto binaryOperationName = ctx->children.at(1)->getText();
    antlrcpp::Any left = visit(ctx->expression(0));
    antlrcpp::Any right = visit(ctx->expression(1));

    Expression* leftExpression1 = left.as<Expression*>();

    std::vector<Expression*> children;
    children.push_back(leftExpression1);
    children.push_back(right.as<Expression*>());

    return MakeExpression(binaryOperationName, children);
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitFuncInvokeExp(ExpressionAntlrParser::FuncInvokeExpContext* context)
{
    ExpressionAntlrParser::ArgsListContext* argContext = context->argsList();
    std::vector<antlr4::tree::ParseTree*> argChildren = argContext->children;

    std::vector<Expression*> childExpressions;
    for (auto argChild : argChildren)
    {
        // TODO - in the C# version this part happens in a ProcessArgsList function that also checks for LambdaContexts
        if (dynamic_cast<ExpressionAntlrParser::ExpressionContext*>(argChild))
        {
            auto childExpression = visit(argChild);
            childExpressions.push_back(childExpression);
        }
    }

    std::string functionName = context->primaryExpression()->getText();
    if (context->NON() != nullptr)
    {
        functionName += context->NON()->getText();
    }

    return MakeExpression(functionName, childExpressions);
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitIdAtom(ExpressionAntlrParser::IdAtomContext* ctx)
{
    return antlrcpp::Any();
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitIndexAccessExp(ExpressionAntlrParser::IndexAccessExpContext* ctx)
{
    return antlrcpp::Any();
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitMemberAccessExp(ExpressionAntlrParser::MemberAccessExpContext* ctx)
{
    return antlrcpp::Any();
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitParenthesisExp(ExpressionAntlrParser::ParenthesisExpContext* ctx)
{
    return antlrcpp::Any();
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitArrayCreationExp(ExpressionAntlrParser::ArrayCreationExpContext* ctx)
{
    return antlrcpp::Any();
}

antlrcpp::Any ExpressionParser::ExpressionTransformer::visitNumericAtom(ExpressionAntlrParser::NumericAtomContext* ctx)
{
    std::string numericString = ctx->getText();

    try
    {
        int integer = std::stoi(numericString);
        return Expression::ConstantExpression(integer);
    }
    catch (const std::bad_any_cast&)
    {
    }

    try
    {
        long long int longInteger = std::stol(numericString);
        return Expression::ConstantExpression(longInteger);
    }
    catch (const std::bad_any_cast&) {}

    try
    {
        double decimalValue = std::stod(numericString);
        return Expression::ConstantExpression(decimalValue);
    }
    catch (const std::bad_any_cast&)
    {
        // throw new Exception($"{context.GetText()} is not a number in expression '{context.GetText()}'");
    }


    return Expression::ConstantExpression(0);
}

/*
public override Expression VisitStringAtom([NotNull] ExpressionAntlrParser.StringAtomContext context)
{

    var text = context.GetText();
    if (text.StartsWith("'", StringComparison.Ordinal) && text.EndsWith("'", StringComparison.Ordinal))
    {
        text = text.Substring(1, text.Length - 2).Replace("\\'", "'");
    }
    else if (text.StartsWith("\"", StringComparison.Ordinal) && text.EndsWith("\"", StringComparison.Ordinal))
    {
        text = text.Substring(1, text.Length - 2).Replace("\\\"", "\"");
    }
    else
    {
        throw new Exception($"Invalid string {text}");
    }

    return Expression.ConstantExpression(EvalEscape(text));

    return nullptr;
}
*/

Expression* ExpressionParser::ExpressionTransformer::MakeExpression(std::string functionType, std::vector<Expression*> children)
{
    return Expression::MakeExpression(m_lookupFunction(functionType), children);
    // If lookup function fails, throw this: throw new SyntaxErrorException($"{functionType} does not have an evaluator, it's not a built-in function or a custom function.")
}

EvaluatorLookup ExpressionParser::getEvaluatorLookup()
{
    return m_evaluatorLookup;
}

/*
Expression* ExpressionParser::Parse(std::string expression)
{

    if (expression.size() == 0)
    {
        return Expression::ConstantExpression(std::string());
    }
    else
    {
        return new ExpressionTransformer(m_evaluatorLookup)->Transform(AntlrParse(expression));
    }

}
*/

ExpressionParser::ExpressionParser(EvaluatorLookup lookup)
{
    m_evaluatorLookup = lookup;
}

antlr4::tree::ParseTree* ExpressionParser::AntlrParse(std::string expression)
{
    /*
    // The expressionDict is a cache i think
    if (expressionDict.TryGetValue(expression, out var expressionParseTree))
    {
        return expressionParseTree;
    }
    */

    auto inputStream = new antlr4::ANTLRInputStream(expression);
    auto lexer = new ExpressionAntlrLexer(inputStream);
    lexer->removeErrorListeners();
    auto tokenStream = new antlr4::CommonTokenStream(lexer);
    auto parser = new ExpressionAntlrParser(tokenStream);
    parser->removeErrorListeners();
    // parser->addErrorListener(new ParserErrorListener());
    parser->setBuildParseTree(true);

    ExpressionAntlrParser::ExpressionContext* expressionContext = nullptr;
    ExpressionAntlrParser::FileContext* fileContext = parser->file();
    if (fileContext != nullptr)
    {
        expressionContext = fileContext->expression();
    }

    // expressionDict.TryAdd(expression, expressionContext);

    return expressionContext;
}

Expression* ExpressionParser::Parse(std::string expression)
{
    if (expression.empty())
    {
        return Expression::ConstantExpression(std::string());
    }
    else
    {
        ExpressionTransformer* transformer = new ExpressionTransformer(m_evaluatorLookup);
        return transformer->Transform(AntlrParse(expression));
    }

}
